NJT binary pointwise broadcasting support via jagged <-> padded dense conversion

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #133021

Related: #132695

This PR uses padded dense <-> jagged conversions to handle binary pointwise broadcasting of (NT, T) and (T, NT). This includes:
* `(B, j0, D) + (1, 1, 1)`
* `(B, j0, D) + (B, 1, 1)`
* `(B, j0, D) + (B, 1, D)`
* etc.

This PR also adds (hacky) support for bool inputs to the jagged <-> padded dense conversions. The underlying CUDA kernels do not support integer / bool inputs; so the following workaround is employed: `convert input -> half, run conversion kernel, convert output -> bool`. Note that this bool support is needed specifically for the backward formula of `fmax`, and likely others.

cc @cpuhrsch @bhosmer @drisspg @soulitzer @davidberard98 @YuqingJ

jbschlosser committed:
NJT binary pointwise broadcasting support via jagged <-> padded dense conversion

[ghstack-poisoned]

jbschlosser committed:
Update on "NJT binary pointwise broadcasting support via jagged <-> padded dense conversion"


Related: #132695

This PR uses padded dense <-> jagged conversions to handle binary pointwise broadcasting of (NT, T) and (T, NT). This includes:
* `(B, j0, D) + (1, 1, 1)`
* `(B, j0, D) + (B, 1, 1)`
* `(B, j0, D) + (B, 1, D)`
* etc.

[ghstack-poisoned]

jbschlosser committed:
Update on "NJT binary pointwise broadcasting support via jagged <-> padded dense conversion"


Related: #132695

This PR uses padded dense <-> jagged conversions to handle binary pointwise broadcasting of (NT, T) and (T, NT). This includes:
* `(B, j0, D) + (1, 1, 1)`
* `(B, j0, D) + (B, 1, 1)`
* `(B, j0, D) + (B, 1, D)`
* etc.

[ghstack-poisoned]

davidberard98 commented on torch/nested/_internal/ops.py:
```diff
@@ -294,14 +294,22 @@ def jagged_binary_pointwise(func, *args, **kwargs):
-        new_values = torch.cat(outputs, dim=0)
-        return NestedTensor(new_values, **extracted_kwargs)
+        # handle broadcasting via jagged -> padded dense conversion
+        padded = nt.to_padded_tensor(0.0)
+
+        # broadcasting function call
+        lhs, rhs = (padded, t) if a_is_nt else (t, padded)
+        output_padded = func(lhs, rhs, *args[2:], **kwargs)
+
+        # padded dense -> jagged
+        total_L = nt._values.shape[0]
+
+        from .nested_tensor import nested_from_padded
+
+        # TODO: propagate cached min / max sequence length
+        return nested_from_padded(
+            output_padded, offsets=nt._offsets, ragged_idx=nt._ragged_idx, sum_S=total_L
+        )
```
Another approach:

```python
t = t.expand([x if i != ragged_idx else max_seqlen for i, x in enumerate(t.shape)])
t_as_nt = padded_dense_to_jagged(t, offsets)
...
lhs, rhs = ...
...
func(lhs, rhs)
```

jbschlosser replied:
True! but do we still get nice fusion properties from Inductor for this approach?

jbschlosser committed:
Update on "NJT binary pointwise broadcasting support via jagged <-> padded dense conversion"


Related: #132695

This PR uses padded dense <-> jagged conversions to handle binary pointwise broadcasting of (NT, T) and (T, NT). This includes:
* `(B, j0, D) + (1, 1, 1)`
* `(B, j0, D) + (B, 1, 1)`
* `(B, j0, D) + (B, 1, D)`
* etc.

[ghstack-poisoned]

jbschlosser committed:
Update on "NJT binary pointwise broadcasting support via jagged <-> padded dense conversion"


Related: #132695

This PR uses padded dense <-> jagged conversions to handle binary pointwise broadcasting of (NT, T) and (T, NT). This includes:
* `(B, j0, D) + (1, 1, 1)`
* `(B, j0, D) + (B, 1, 1)`
* `(B, j0, D) + (B, 1, D)`
* etc.

[ghstack-poisoned]

jbschlosser committed:
Update on "NJT binary pointwise broadcasting support via jagged <-> padded dense conversion"


Related: #132695

This PR uses padded dense <-> jagged conversions to handle binary pointwise broadcasting of (NT, T) and (T, NT). This includes:
* `(B, j0, D) + (1, 1, 1)`
* `(B, j0, D) + (B, 1, 1)`
* `(B, j0, D) + (B, 1, D)`
* etc.

[ghstack-poisoned]

jbschlosser committed:
Update on "NJT binary pointwise broadcasting support via jagged <-> padded dense conversion"


Related: #132695

This PR uses padded dense <-> jagged conversions to handle binary pointwise broadcasting of (NT, T) and (T, NT). This includes:
* `(B, j0, D) + (1, 1, 1)`
* `(B, j0, D) + (B, 1, 1)`
* `(B, j0, D) + (B, 1, D)`
* etc.

[ghstack-poisoned]

jbschlosser committed:
Update on "NJT binary pointwise broadcasting support via jagged <-> padded dense conversion"


Related: #132695

This PR uses padded dense <-> jagged conversions to handle binary pointwise broadcasting of (NT, T) and (T, NT). This includes:
* `(B, j0, D) + (1, 1, 1)`
* `(B, j0, D) + (B, 1, 1)`
* `(B, j0, D) + (B, 1, D)`
* etc.

cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ

[ghstack-poisoned]

jbschlosser committed:
Update on "NJT binary pointwise broadcasting support via jagged <-> padded dense conversion"


Related: #132695

This PR uses padded dense <-> jagged conversions to handle binary pointwise broadcasting of (NT, T) and (T, NT). This includes:
* `(B, j0, D) + (1, 1, 1)`
* `(B, j0, D) + (B, 1, 1)`
* `(B, j0, D) + (B, 1, D)`
* etc.

cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ

[ghstack-poisoned]

jbschlosser committed:
Update on "NJT binary pointwise broadcasting support via jagged <-> padded dense conversion"


Related: #132695

This PR uses padded dense <-> jagged conversions to handle binary pointwise broadcasting of (NT, T) and (T, NT). This includes:
* `(B, j0, D) + (1, 1, 1)`
* `(B, j0, D) + (B, 1, 1)`
* `(B, j0, D) + (B, 1, D)`
* etc.

cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ

[ghstack-poisoned]

jbschlosser committed:
Update on "NJT binary pointwise broadcasting support via jagged <-> padded dense conversion"


Related: #132695

This PR uses padded dense <-> jagged conversions to handle binary pointwise broadcasting of (NT, T) and (T, NT). This includes:
* `(B, j0, D) + (1, 1, 1)`
* `(B, j0, D) + (B, 1, 1)`
* `(B, j0, D) + (B, 1, D)`
* etc.

cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ

[ghstack-poisoned]

jbschlosser committed:
Update on "NJT binary pointwise broadcasting support via jagged <-> padded dense conversion"


Related: #132695

This PR uses padded dense <-> jagged conversions to handle binary pointwise broadcasting of (NT, T) and (T, NT). This includes:
* `(B, j0, D) + (1, 1, 1)`
* `(B, j0, D) + (B, 1, 1)`
* `(B, j0, D) + (B, 1, D)`
* etc.

cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ

[ghstack-poisoned]

jbschlosser committed:
Update on "NJT binary pointwise broadcasting support via jagged <-> padded dense conversion"


Related: #132695

This PR uses padded dense <-> jagged conversions to handle binary pointwise broadcasting of (NT, T) and (T, NT). This includes:
* `(B, j0, D) + (1, 1, 1)`
* `(B, j0, D) + (B, 1, 1)`
* `(B, j0, D) + (B, 1, D)`
* etc.

cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ

[ghstack-poisoned]

jbschlosser committed:
Update on "NJT binary pointwise broadcasting support via jagged <-> padded dense conversion"


Related: #132695

This PR uses padded dense <-> jagged conversions to handle binary pointwise broadcasting of (NT, T) and (T, NT). This includes:
* `(B, j0, D) + (1, 1, 1)`
* `(B, j0, D) + (B, 1, 1)`
* `(B, j0, D) + (B, 1, D)`
* etc.

cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ

[ghstack-poisoned]

jbschlosser committed:
Update on "NJT binary pointwise broadcasting support via jagged <-> padded dense conversion"


Related: #132695

This PR uses padded dense <-> jagged conversions to handle binary pointwise broadcasting of (NT, T) and (T, NT). This includes:
* `(B, j0, D) + (1, 1, 1)`
* `(B, j0, D) + (B, 1, 1)`
* `(B, j0, D) + (B, 1, D)`
* etc.

cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ

[ghstack-poisoned]

jbschlosser committed:
Update on "NJT binary pointwise broadcasting support via jagged <-> padded dense conversion"


Related: #132695

This PR uses padded dense <-> jagged conversions to handle binary pointwise broadcasting of (NT, T) and (T, NT). This includes:
* `(B, j0, D) + (1, 1, 1)`
* `(B, j0, D) + (B, 1, 1)`
* `(B, j0, D) + (B, 1, D)`
* etc.

cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ

[ghstack-poisoned]

jbschlosser committed:
Update on "NJT binary pointwise broadcasting support via jagged <-> padded dense conversion"


Related: #132695

This PR uses padded dense <-> jagged conversions to handle binary pointwise broadcasting of (NT, T) and (T, NT). This includes:
* `(B, j0, D) + (1, 1, 1)`
* `(B, j0, D) + (B, 1, 1)`
* `(B, j0, D) + (B, 1, D)`
* etc.

cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ

[ghstack-poisoned]

jbschlosser committed:
Update on "NJT binary pointwise broadcasting support via jagged <-> padded dense conversion"


Related: #132695

This PR uses padded dense <-> jagged conversions to handle binary pointwise broadcasting of (NT, T) and (T, NT). This includes:
* `(B, j0, D) + (1, 1, 1)`
* `(B, j0, D) + (B, 1, 1)`
* `(B, j0, D) + (B, 1, D)`
* etc.

cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ

[ghstack-poisoned]

jbschlosser committed:
Update on "NJT binary pointwise broadcasting support via jagged <-> padded dense conversion"


Related: #132695

This PR uses padded dense <-> jagged conversions to handle binary pointwise broadcasting of (NT, T) and (T, NT). This includes:
* `(B, j0, D) + (1, 1, 1)`
* `(B, j0, D) + (B, 1, 1)`
* `(B, j0, D) + (B, 1, D)`
* etc.

cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ

[ghstack-poisoned]

jbschlosser committed:
Update on "NJT binary pointwise broadcasting support via jagged <-> padded dense conversion"


Related: #132695

This PR uses padded dense <-> jagged conversions to handle binary pointwise broadcasting of (NT, T) and (T, NT). This includes:
* `(B, j0, D) + (1, 1, 1)`
* `(B, j0, D) + (B, 1, 1)`
* `(B, j0, D) + (B, 1, D)`
* etc.

cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ

[ghstack-poisoned]

jbschlosser committed:
Update on "NJT binary pointwise broadcasting support via jagged <-> padded dense conversion"


Related: #132695

This PR uses padded dense <-> jagged conversions to handle binary pointwise broadcasting of (NT, T) and (T, NT). This includes:
* `(B, j0, D) + (1, 1, 1)`
* `(B, j0, D) + (B, 1, 1)`
* `(B, j0, D) + (B, 1, D)`
* etc.

cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ

[ghstack-poisoned]

jbschlosser committed:
Update on "NJT binary pointwise broadcasting support via jagged <-> padded dense conversion"


Related: #132695

This PR uses padded dense <-> jagged conversions to handle binary pointwise broadcasting of (NT, T) and (T, NT). This includes:
* `(B, j0, D) + (1, 1, 1)`
* `(B, j0, D) + (B, 1, 1)`
* `(B, j0, D) + (B, 1, D)`
* etc.

cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ

[ghstack-poisoned]

jbschlosser committed:
Update on "NJT binary pointwise broadcasting support via jagged <-> padded dense conversion"


Related: #132695

This PR uses padded dense <-> jagged conversions to handle binary pointwise broadcasting of (NT, T) and (T, NT). This includes:
* `(B, j0, D) + (1, 1, 1)`
* `(B, j0, D) + (B, 1, 1)`
* `(B, j0, D) + (B, 1, D)`
* etc.

cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ

[ghstack-poisoned]

jbschlosser committed:
Update on "NJT binary pointwise broadcasting support via jagged <-> padded dense conversion"


Related: #132695

This PR uses padded dense <-> jagged conversions to handle binary pointwise broadcasting of (NT, T) and (T, NT). This includes:
* `(B, j0, D) + (1, 1, 1)`
* `(B, j0, D) + (B, 1, 1)`
* `(B, j0, D) + (B, 1, D)`
* etc.

cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ

[ghstack-poisoned]

jbschlosser committed:
Update on "NJT binary pointwise broadcasting support via jagged <-> padded dense conversion"


Related: #132695

This PR uses padded dense <-> jagged conversions to handle binary pointwise broadcasting of (NT, T) and (T, NT). This includes:
* `(B, j0, D) + (1, 1, 1)`
* `(B, j0, D) + (B, 1, 1)`
* `(B, j0, D) + (B, 1, D)`
* etc.

cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ

[ghstack-poisoned]

jbschlosser committed:
Update on "NJT binary pointwise broadcasting support via jagged <-> padded dense conversion"


Related: #132695

This PR uses padded dense <-> jagged conversions to handle binary pointwise broadcasting of (NT, T) and (T, NT). This includes:
* `(B, j0, D) + (1, 1, 1)`
* `(B, j0, D) + (B, 1, 1)`
* `(B, j0, D) + (B, 1, D)`
* etc.

cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ

[ghstack-poisoned]

jbschlosser committed:
Update on "NJT binary pointwise broadcasting support via jagged <-> padded dense conversion"


Related: #132695

This PR uses padded dense <-> jagged conversions to handle binary pointwise broadcasting of (NT, T) and (T, NT). This includes:
* `(B, j0, D) + (1, 1, 1)`
* `(B, j0, D) + (B, 1, 1)`
* `(B, j0, D) + (B, 1, D)`
* etc.

cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ

[ghstack-poisoned]

jbschlosser committed:
Update on "NJT binary pointwise broadcasting support via jagged <-> padded dense conversion"


Related: #132695

This PR uses padded dense <-> jagged conversions to handle binary pointwise broadcasting of (NT, T) and (T, NT). This includes:
* `(B, j0, D) + (1, 1, 1)`
* `(B, j0, D) + (B, 1, 1)`
* `(B, j0, D) + (B, 1, D)`
* etc.

This PR also adds (hacky) support for bool inputs to the jagged <-> padded dense conversions. The underlying CUDA kernels do not support integer / bool inputs; so the following workaround is employed: `convert input -> half, run conversion kernel, convert output -> bool`. Note that this bool support is needed specifically for the backward formula of `fmax`, and likely others.

cc cpuhrsch bhosmer drisspg soulitzer davidberard98 YuqingJ

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge
